ql/indexes/equityindex.hpp

```diff
@@ -0,0 +1,119 @@
+    /*! The equity index object allows to retrieve past fixings,
+        as well as project future fixings using either both
+        the risk free interest rate term structure and the dividend
+        term structure, or just the interest rate term structure
+        in which case one can provide a term structure of equity
+        forwards implied from, e.g. option prices.
+        
+        In case of the first method, the forward is calculated as:
+        \f[
+        I(t, T) = I(t, t) \frac{P_{D}(t, T)}{P_{R}(t, T)},
+        \f]
+        where \f$ I(t, t) \f$ is today's value of the index,
+        \f$ P_{D}(t, T) \f$ is a discount factor of the dividend
+        curve at future time \f$ T \f$, and \f$ P_{R}(t, T) \f$ is
+        a discount factor of the risk free curve at future time
+        \f$ T \f$.
+
+        In case of the latter method, the forward is calculated as:
+        \f[
+        I(t, T) = I(t, t) \frac{1}{P_{F}(t, T)},
+        \f]
+        where \f$ P_{F}(t, T) \f$ is a discount factor of the equity
+        forward term structure.
+
+        To forecast future fixings, the index requires today's fixing.
+    */
+    class EquityIndex : public Index, public Observer {
+      public:
+        EquityIndex(std::string name,
+                    Currency currency,
+                    Calendar fixingCalendar,
+                    Handle<YieldTermStructure> interest = {},
+                    Handle<YieldTermStructure> dividend = {});
+
+        //! \name Index interface
+        //@{
+        std::string name() const override { return name_; }
+        Calendar fixingCalendar() const override { return settlementCalendar_; }
+        bool isValidFixingDate(const Date& fixingDate) const override;
+        Real fixing(const Date& fixingDate, bool forecastTodaysFixing = false) const override;
+        //@}
+        //! \name Observer interface
+        //@{
+        void update() override;
+        //@}
+        //! \name Inspectors
+        //@{
+        const Currency& currency() const { return currency_; }
+        //! the rate curve used to forecast fixings
+        Handle<YieldTermStructure> equityInterestRateCurve() const { return interest_; }
+        //! the dividend curve used to forecast fixings
+        Handle<YieldTermStructure> equityDividendCurve() const { return dividend_; }
+        //@}
+        //! \name Fixing calculations
+        //@{
+        //! It can be overridden to implement particular conventions
+        virtual Real forecastFixing(const Date& fixingDate) const;
+        virtual Real pastFixing(const Date& fixingDate) const;
+        // @}
+        //! \name Other methods
+        //@{
+        //! returns a copy of itself linked to a different forwarding curve
+        virtual ext::shared_ptr<EquityIndex> clone(
+                const Handle<YieldTermStructure>& interest,
+                const Handle<YieldTermStructure>& dividend) const;
+        // @}
+
+      protected:
```

Comment:
I would keep these private until we can't.  Derived classes can set them from the constructor and read them from methods.

Reply:
Ok, I will keep them private for now.